app/src/main/java/com/example/chessclock/feature/clock/ClockViewModel.kt

```diff
@@ -0,0 +1,216 @@
+package com.example.chessclock.feature.clock
+
+import androidx.hilt.Assisted
+import androidx.hilt.lifecycle.ViewModelInject
+import androidx.lifecycle.*
+import kotlinx.coroutines.*
+import kotlinx.coroutines.channels.ReceiveChannel
+import kotlinx.coroutines.channels.ticker
+import java.lang.Math.random
+
+sealed class Player(var millisLeft: Float, private val incrementMillis: Int = 0) {
+    private val seconds: Int
+        get() = millisLeft.toInt() / 1000 % 60
+    private val minutes: Int
+        get() = millisLeft.toInt() / 1000 / 60
+    private val hours: Int
+        get() = millisLeft.toInt() / 1000 / 60 / 60
+
+    open val text: String
+        get() {
+            val hoursText = if (hours > 0) {
+                "${formatTime(hours)}:"
+            } else {
+                ""
+            }
+            return "$hoursText${formatTime(minutes)}:${formatTime(seconds)}"
+        }
+
+
+    fun incrementTime() {
+        millisLeft += incrementMillis
+    }
+
+    val hasLost: Boolean
+        get() = millisLeft <= 0
+
+    private fun formatTime(time: Int): String =
+        if (time < 10) {
+            "0$time"
+        } else {
+            time.toString()
+        }
+
+    override fun toString(): String {
+        return "Player(millisLeft=$millisLeft, increment=$incrementMillis)"
+    }
+
+    class White(millisLeft: Float, increment: Int = 0) : Player(millisLeft, increment) {
+        override val text: String
+            get() = if (hasLost) "Black wins" else super.text
+    }
+
+    class Black(millisLeft: Float, increment: Int = 0) : Player(millisLeft, increment) {
+        override val text: String
+            get() = if (hasLost) "White wins" else super.text
+    }
+}
+
+sealed class PlayerDisplay(
```

Comment:
add comment why it is different class - I guess it was because using Player would spam too many events, test if it still is the case?